module4/lessons/jest_supertest_and_graphql.md

```diff
@@ -0,0 +1,213 @@
+      "dialect": "postgres",
+      "seederStorage": "sequelize"
+    }
+}
+```
+Now that we have a test configuration in our `database.json`, let's create the database:
+```bash
+$ psql
+psql (13.1)
+Type "help" for help.
+
+<YOUR_NAME_OR_POSTGRES_USER_NAME>=# CREATE DATABASE crate_testing;
+CREATE DATABASE
+\l; # <= Check to make sure the DB was created
+<A LIST OF DATABASES>
+\q # <= Quit psql
+```
+Great! Now we have a test db, so let's add some data to it:
+
+```bash
+$ NODE_ENV=test npm setup:db
+```
+If we look at our `package.json` file, we will notice a `scripts` block. You can think of these like `Rake` commands. So, anything in here we can run via `npm`. Now that we have migrated our test db and added some data to it, we should make sure it worked:
+```bash
+$ psql crate_testing
+<crate_testing>=# SELECT * FROM USERS;
+<...SOME USER DATA...>
+```
+Great, we now have a test database and seeded some data to use in our tests. Note: This may be a slightly different workflow than say a Rails App, where we create/destroy data per test, but for now, this will get us up and running!
+
+
+## Setting up Jest and Supertest
+In Rails, we have RSpec--or less often, MiniTest--, in Node we have...lots of options; for this lesson, we are going to use Jest for our testing library, as well as Supertest, which allows us to mock our server. So, let's get started.
+
+```bash
+$ npm install --save-dev jest
+$ npm install --save-dev supertest
+```
+Here, the `save-dev` flag puts these libraries into our development dependencies; that is, they won't be compiled and added to our production JS build. We can see this in our `package.json` file:
+```js
+ ...,
+  "devDependencies": {
+    "@babel/cli": "7.11.6",
+    "@babel/core": "7.11.6",
+    "@babel/node": "7.10.5",
+    "@babel/plugin-transform-runtime": "7.11.5",
+    "@babel/preset-env": "7.11.5",
+    "jest": "^26.6.3",
+    "nodemon": "2.0.4",
+    "supertest": "^6.0.1"
+  },
+ ...
+```
+Alright, now we have our testing libraries installed, so let's write a test! In Node applications, you will traditionally see specs nested within the actual code directories themselves, as opposed to a seperated directory like `spec` in Rails. This is all up to you on how you want to structure your spec files, but for this exercise, lets add them directly to our code. Our file structure will end up looking like this:
+```
+ code
+  | api
+    | src
+      | modules
+        | <module_name>
+          | tests
+            | our tests
+```
+Jest will look for any file that ends with `.test.js`, so we can begin with our User model and test there.
+
+```bash
+ $ cd code/api/src/modules/user
+ $ mkdir tests
+ $ cd tests
+ $ touch query.test.js
+```
+Open that file and we will setup Jest and Supertest with our Mock Server.
+
+In order for Supertest to connect to our `graphql` schema, we will need to set up a mock server. If we look in `code/api/src/setup/graphql.js`, we will see how we are setting up our schema with express:
```

Comment:
I think a small note here about why the schema needs to be connected to the mock server would be helpful. I could see it being confusing since there is already a dev server that is connected to the DB, why do we have to do this explicitly with the text mock server?

Reply:
Good call